### ISSUE TITLE
Batch stuck in pending with JIDs showing as both pending and failed

### ISSUE THREAD

**Issue by mmarinaccio** [sidekiq/sidekiq#6260]
Ruby version: 3.2.2p53
Rails version: 6.1.7.7
Sidekiq / Pro / Enterprise version(s): sidekiq-pro (7.2.0)

**config/sidekiq.yml**
```yml
# Options here can still be overridden by cmd line args.
---
:verbose: false
:concurrency: 10
:timeout: 25

production:
  :concurrency: 10
sandbox:
  :concurrency: 10
test:
  :concurrency: 1
development:
  :concurrency: 5
```

**config/initializers/sidekiq.rb**
```ruby
Sidekiq.configure_server do |config|
  config.redis = {
    host: ENV.fetch('SIDEKIQ_REDIS_HOST', 'localhost'),
    port: ENV.fetch('SIDEKIQ_REDIS_PORT', 6379),
  }

  config.dogstatsd = -> { Datadog::Statsd.new('localhost', 8125, namespace: 'sidekiq') }

  config.server_middleware do |chain|
    require 'sidekiq/middleware/server/statsd'

    chain.add(Sidekiq::Middleware::Server::Statsd)

    chain.add(Sidekiq::CustomMiddleware::ReleaseLock)
  end

  config.logger.level = Logger::INFO
end

Sidekiq.configure_client do |config|
  config.redis = {
    host: ENV.fetch('SIDEKIQ_REDIS_HOST', 'localhost'),
    port: ENV.fetch('SIDEKIQ_REDIS_PORT', 6379),
  }
end

Sidekiq.default_job_options = { backtrace: false, retry: 0 }
```
## Issue
I've observed a few batches that appear to be stuck in a _pending_ state. What's confusing to me is that the same JID is present in both the `Pending JIDS` and `Dead JIDS` arrays, as shown below. These jobs don't have retry policies - i.e. I have retry set to 0 in the initializer and no Sidekiq jobs override this setting.

![image](https://github.com/sidekiq/sidekiq/assets/20650053/cc18bfb8-04ff-45ca-bd9a-e8c60b4b3a6e)

The below logs show that one of these jobs failed:
```
"@timestamp": "2024-04-20 02:52:19.962",
"@message": "2024-04-20T02:52:19.962Z pid=29 tid=18it WARN: {\"context\":\"Job raised exception\",\"job\":{\"retry\":0,\"queue\":\"days_crd\",\"backtrace\":false,\"class\":\"Obfuscated::OnboardObfuscatedUserJob\",\"args\":[{\"studio_id\":5309,\"external_id\":94799925}],\"jid\":\"f1231d9021fe81db62bfdb73\",\"created_at\":1713580637.0469265,\"bid\":\"DkISOnegRJ2xqQ\",\"enqueued_at\":1713580637.0899472}}",

"@timestamp": "2024-04-20 02:52:19.961",
"@message": "2024-04-20T02:52:19.961Z pid=29 tid=18it class=Obfuscated::OnboardObfuscatedUserJob jid=f1231d9021fe81db62bfdb73 bid=DkISOnegRJ2xqQ elapsed=0.45 INFO: fail",

"@timestamp": "2024-04-20 02:52:19.956",
"@message": "2024-04-20T02:52:19.956Z pid=29 tid=18it class=Obfuscated::OnboardObfuscatedUserJob jid=f1231d9021fe81db62bfdb73 bid=DkISOnegRJ2xqQ INFO: Adding dead Obfuscated::OnboardObfuscatedUserJob job f1231d9021fe81db62bfdb73",

"@timestamp": "2024-04-20 02:52:19.511",
"@message": "2024-04-20T02:52:19.511Z pid=29 tid=18it class=Obfuscated::OnboardObfuscatedUserJob jid=f1231d9021fe81db62bfdb73 bid=DkISOnegRJ2xqQ INFO: start",
```

I'd expect this to increment the Failures count and pop the JID out of pending.

TLDR: why do my cleanly failing jobs in a batch exist as both failed and pending jobs?

**Comment by mperham** [comment]
Because a job is pending until it succeeds. Failure can be transient. Even a dead job can be manually retried and succeed.

**Comment by mmarinaccio** [comment]
I appreciate your prompt response, @mperham. Got it, that makes sense.

What I failed to mention is that my `on_complete` callback is failing to send, since `status.complete? == false`. Sounds like this is a separate issue caused by failing to actually execute all jobs enqueued in the context of a batch.

Would you be able to point me in the right direction for debugging this? Seems odd that my system is somehow registering a job with a batch but never actually executing the job. If it helps limit the possibilities, my batch creation looks like this:
```ruby
sidekiq_batch = Sidekiq::Batch.new

sidekiq_batch.description = 'Example batch'

sidekiq_batch.on(:complete, Callbacks::MyCallback, 'studio_id' => @studio.id)

@api_client.get_all_users(@studio.studio_integration_id).each do |users|
  sidekiq_batch.jobs do
    Sidekiq::Client.push_bulk(
      'class' => JobRunInBatch,
      'args'  => users.map do |user|
        [
          {
            'studio_id'   => @studio.id,
            'external_id' => user['UserID'],
          },
        ]
      end
    )
  end
end
```

Otherwise, I'll close this out and re-open a more appropriate issue if I don't make any progress.

Thanks again.

**Comment by mperham** [comment]
This (stuck batches with Sidekiq 7+) has happened to a few customers but I've been unable to reproduce it or track down the cause. Make sure you've updated everything as much as possible: Sidekiq, Redis and associated libraries.

**Comment by mmarinaccio** [comment]
Ah @mperham I figured it out. I was calling `batch.jobs` more than once, which you explicitly describe as unsafe ([link](https://github.com/sidekiq/sidekiq/wiki/Batches#adding-jobs)).

The fix was to turn this:
```ruby
@api_client.get_all_users(@studio.studio_integration_id).each do |users|
  sidekiq_batch.jobs do
  #...
```

Into this:
```ruby
sidekiq_batch.jobs do
  @api_client.get_all_users(@studio.studio_integration_id).each do |crd_user_batch
  # ...
```

Thanks for your fine work on Sidekiq. It's a real joy to use.
